File: templates/scenarios/react-vite/api/internal/testutil/modeltest/spec.go
package modeltest

import (
	"encoding/json"
	"fmt"
	"os"
)

// WorkflowSpec is a declarative workflow contract used by tests to keep
// hand-authored matrices and traces aligned with the documented model.
type WorkflowSpec struct {
	ID             string               `json:"id"`
	Domain         string               `json:"domain"`
	Description    string               `json:"description"`
	States         []string             `json:"states"`
	Events         []string             `json:"events"`
	InitialState   string               `json:"initialState"`
	TerminalStates []string             `json:"terminalStates"`
	Transitions    []SpecTransition     `json:"transitions"`
	Invariants     []string             `json:"invariants"`
	Traces         []SpecTrace          `json:"traces"`
	FormalModel    *SpecFormalModelInfo `json:"formalModel,omitempty"`
}

type SpecTransition struct {
	From      string `json:"from"`
	Event     string `json:"event"`
	To        string `json:"to"`
	WantError bool   `json:"wantError,omitempty"`
}

type SpecTrace struct {
	Name    string          `json:"name"`
	Initial string          `json:"initial"`
	Steps   []SpecTraceStep `json:"steps"`
}

type SpecTraceStep struct {
	Event     string `json:"event"`
	Want      string `json:"want"`
	WantError bool   `json:"wantError,omitempty"`
}

type SpecFormalModelInfo struct {
	Status             string `json:"status"`
	Tool               string `json:"tool,omitempty"`
	GeneratedArtifacts string `json:"generatedArtifacts,omitempty"`
	DriftCheck         string `json:"driftCheck,omitempty"`
}

// LoadWorkflowSpec reads a JSON workflow spec from path.
func LoadWorkflowSpec(t TestingT, path string) WorkflowSpec {
	t.Helper()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read workflow spec %s: %v", path, err)
	}
	var spec WorkflowSpec
	if err := json.Unmarshal(data, &spec); err != nil {
		t.Fatalf("parse workflow spec %s: %v", path, err)
	}
	return spec
}

// ValidateWorkflowSpecConformance verifies that a declarative workflow spec
// agrees with the production status/event lists plus matrix and trace tests.
func ValidateWorkflowSpecConformance[S comparable, E comparable](
	spec WorkflowSpec,
	statuses []S,
	events []E,
	rows []MatrixRow[S, E],
	traces []Trace[S, E],
) []error {
	var errs []error
	if spec.ID == "" {
		errs = append(errs, fmt.Errorf("spec id is required"))
	}
	if spec.Domain == "" {
		errs = append(errs, fmt.Errorf("spec domain is required"))
	}
	if spec.InitialState == "" {
		errs = append(errs, fmt.Errorf("spec initialState is required"))
	}
	if len(spec.Transitions) == 0 {
		errs = append(errs, fmt.Errorf("spec transitions must not be empty"))
	}

	specStates := stringSet(spec.States)
	for _, status := range statuses {
		value := fmt.Sprint(status)
		if !specStates[value] {
			errs = append(errs, fmt.Errorf("spec missing production state %s", value))
		}
	}
	for _, state := range spec.States {
		if !containsValue(statuses, state) {
			errs = append(errs, fmt.Errorf("spec state %s is not a production state", state))
		}
	}
	if !specStates[spec.InitialState] {
		errs = append(errs, fmt.Errorf("spec initialState %s is not a known state", spec.InitialState))
	}
	for _, terminal := range spec.TerminalStates {
		if !specStates[terminal] {
			errs = append(errs, fmt.Errorf("spec terminal state %s is not a known state", terminal))
		}
	}

	specEvents := stringSet(spec.Events)
	for _, event := range events {
		value := fmt.Sprint(event)
		if !specEvents[value] {
			errs = append(errs, fmt.Errorf("spec missing production event %s", value))
		}
	}
	for _, event := range spec.Events {
		if !containsValue(events, event) {
			errs = append(errs, fmt.Errorf("spec event %s is not a production event", event))
		}
	}

	matrixByPair := make(map[string]SpecTransition, len(rows))
	for _, row := range rows {
		matrixByPair[pairString(fmt.Sprint(row.From), fmt.Sprint(row.Event))] = SpecTransition{
			From:      fmt.Sprint(row.From),
			Event:     fmt.Sprint(row.Event),
			To:        fmt.Sprint(row.To),
			WantError: row.WantErr,
		}
	}
	specByPair := make(map[string]SpecTransition, len(spec.Transitions))
	for _, transition := range spec.Transitions {
		key := pairString(transition.From, transition.Event)
		if _, ok := specByPair[key]; ok {
			errs = append(errs, fmt.Errorf("spec duplicate transition %s/%s", transition.From, transition.Event))
			continue
		}
		specByPair[key] = transition
		row, ok := matrixByPair[key]
		if !ok {
			errs = append(errs, fmt.Errorf("spec transition %s/%s missing from matrix", transition.From, transition.Event))
			continue
		}
		if transition.To != row.To || transition.WantError != row.WantError {
			errs = append(errs, fmt.Errorf(
				"spec transition %s/%s mismatch: spec to=%s wantError=%v matrix to=%s wantError=%v",
				transition.From,
				transition.Event,
				transition.To,
				transition.WantError,
				row.To,
				row.WantError,
			))
		}
	}
	for key, row := range matrixByPair {
		if _, ok := specByPair[key]; !ok {
			errs = append(errs, fmt.Errorf("matrix transition %s/%s missing from spec", row.From, row.Event))
		}
	}

	traceByName := make(map[string]Trace[S, E], len(traces))
	for i, trace := range traces {
		name := trace.Name
		if name == "" {
			name = fmt.Sprintf("trace %d", i)
		}
		traceByName[name] = trace
	}
	for _, specTrace := range spec.Traces {
		trace, ok := traceByName[specTrace.Name]
		if !ok {
			errs = append(errs, fmt.Errorf("spec trace %s missing from tests", specTrace.Name))
			continue
		}
		if specTrace.Initial != fmt.Sprint(trace.Initial) {
			errs = append(errs, fmt.Errorf("spec trace %s initial=%s test initial=%v", specTrace.Name, specTrace.Initial, trace.Initial))
		}
		if len(specTrace.Steps) != len(trace.Steps) {
			errs = append(errs, fmt.Errorf("spec trace %s step count=%d test step count=%d", specTrace.Name, len(specTrace.Steps), len(trace.Steps)))
			continue
		}
		for i, specStep := range specTrace.Steps {
			testStep := trace.Steps[i]
			if specStep.Event != fmt.Sprint(testStep.Event) ||
				specStep.Want != fmt.Sprint(testStep.Want) ||
				specStep.WantError != testStep.WantErr {
				errs = append(errs, fmt.Errorf("spec trace %s step %d differs from test trace", specTrace.Name, i))
			}
		}
	}
	return errs
}

// AssertWorkflowSpecConformance fails t when the spec and executable tests drift.
func AssertWorkflowSpecConformance[S comparable, E comparable](
	t TestingT,
	spec WorkflowSpec,
	statuses []S,
	events []E,
	rows []MatrixRow[S, E],
	traces []Trace[S, E],
) {
	t.Helper()
	if errs := ValidateWorkflowSpecConformance(spec, statuses, events, rows, traces); len(errs) > 0 {
		t.Fatalf("workflow spec mismatch:\n%s", formatErrors(errs))
	}
}

func containsValue[T comparable](values []T, target string) bool {
	for _, value := range values {
		if fmt.Sprint(value) == target {
			return true
		}
	}
	return false
}

func pairString(state string, event string) string {
	return state + "\x00" + event
}

func stringSet(values []string) map[string]bool {
	set := make(map[string]bool, len(values))
	for _, value := range values {
		set[value] = true
	}
	return set
}

File: templates/scenarios/react-vite/ui/src/test-utils/modeltest/spec.ts
import type { MatrixRow } from "./matrix";
import type { Trace } from "./traces";

export interface WorkflowSpec {
  readonly id: string;
  readonly domain: string;
  readonly description: string;
  readonly states: readonly string[];
  readonly events: readonly string[];
  readonly initialState: string;
  readonly terminalStates: readonly string[];
  readonly transitions: readonly SpecTransition[];
  readonly invariants: readonly string[];
  readonly traces: readonly SpecTrace[];
  readonly formalModel?: {
    readonly status: string;
    readonly tool?: string;
    readonly model?: string;
    readonly generatedArtifacts?: string;
    readonly driftCheck?: string;
  };
}

export interface SpecTransition {
  readonly from: string;
  readonly event: string;
  readonly to: string;
  readonly wantError?: boolean;
}

export interface SpecTrace {
  readonly name: string;
  readonly initial: string;
  readonly steps: readonly SpecTraceStep[];
}

export interface SpecTraceStep {
  readonly event: string;
  readonly want: string;
  readonly wantError?: boolean;
}

const pairKey = (state: PropertyKey, event: PropertyKey) => `${String(state)}\u0000${String(event)}`;

export const validateWorkflowSpecConformance = <State extends PropertyKey, Event extends PropertyKey>(
  spec: WorkflowSpec,
  states: readonly State[],
  events: readonly Event[],
  rows: readonly MatrixRow<State, Event>[],
  traces: readonly Trace<State, Event>[],
): string[] => {
  const errors: string[] = [];

  if (spec.id.trim() === "") {
    errors.push("spec id is required");
  }
  if (spec.domain.trim() === "") {
    errors.push("spec domain is required");
  }
  if (spec.initialState.trim() === "") {
    errors.push("spec initialState is required");
  }
  if (spec.transitions.length === 0) {
    errors.push("spec transitions must not be empty");
  }

  const specStates = new Set(spec.states);
  states.forEach((state) => {
    if (!specStates.has(String(state))) {
      errors.push(`spec missing production state ${String(state)}`);
    }
  });
  spec.states.forEach((state) => {
    if (!states.map(String).includes(state)) {
      errors.push(`spec state ${state} is not a production state`);
    }
  });
  if (!specStates.has(spec.initialState)) {
    errors.push(`spec initialState ${spec.initialState} is not a known state`);
  }
  spec.terminalStates.forEach((state) => {
    if (!specStates.has(state)) {
      errors.push(`spec terminal state ${state} is not a known state`);
    }
  });

  const specEvents = new Set(spec.events);
  events.forEach((event) => {
    if (!specEvents.has(String(event))) {
      errors.push(`spec missing production event ${String(event)}`);
    }
  });
  spec.events.forEach((event) => {
    if (!events.map(String).includes(event)) {
      errors.push(`spec event ${event} is not a production event`);
    }
  });

  const matrixByPair = new Map<string, SpecTransition>();
  rows.forEach((row) => {
    matrixByPair.set(pairKey(row.from, row.event), {
      from: String(row.from),
      event: String(row.event),
      to: String(row.to),
      wantError: row.wantError,
    });
  });

  const specByPair = new Map<string, SpecTransition>();
  spec.transitions.forEach((transition) => {
    const key = pairKey(transition.from, transition.event);
    if (specByPair.has(key)) {
      errors.push(`spec duplicate transition ${transition.from}/${transition.event}`);
      return;
    }
    specByPair.set(key, transition);
    const row = matrixByPair.get(key);
    if (!row) {
      errors.push(`spec transition ${transition.from}/${transition.event} missing from matrix`);
      return;
    }
    const specWantError = transition.wantError ?? false;
    const rowWantError = row.wantError ?? false;
    if (transition.to !== row.to || specWantError !== rowWantError) {
      errors.push(
        `spec transition ${transition.from}/${transition.event} mismatch: spec to=${transition.to} wantError=${specWantError} matrix to=${row.to} wantError=${rowWantError}`,
      );
    }
  });

  matrixByPair.forEach((row, key) => {
    if (!specByPair.has(key)) {
      errors.push(`matrix transition ${row.from}/${row.event} missing from spec`);
    }
  });

  const traceByName = new Map<string, Trace<State, Event>>();
  traces.forEach((trace, index) => {
    traceByName.set(trace.name ?? `trace ${index}`, trace);
  });
  spec.traces.forEach((specTrace) => {
    const trace = traceByName.get(specTrace.name);
    if (!trace) {
      errors.push(`spec trace ${specTrace.name} missing from tests`);
      return;
    }
    if (specTrace.initial !== String(trace.initial)) {
      errors.push(`spec trace ${specTrace.name} initial=${specTrace.initial} test initial=${String(trace.initial)}`);
    }
    if (specTrace.steps.length !== trace.steps.length) {
      errors.push(`spec trace ${specTrace.name} step count=${specTrace.steps.length} test step count=${trace.steps.length}`);
      return;
    }
    specTrace.steps.forEach((specStep, index) => {
      const testStep = trace.steps[index];
      if (!testStep) {
        errors.push(`spec trace ${specTrace.name} step ${index} missing from test trace`);
        return;
      }
      if (
        specStep.event !== String(testStep.event) ||
        specStep.want !== String(testStep.want) ||
        (specStep.wantError ?? false) !== (testStep.wantError ?? false)
      ) {
        errors.push(`spec trace ${specTrace.name} step ${index} differs from test trace`);
      }
    });
  });

  return errors;
};

export const assertWorkflowSpecConformance = <State extends PropertyKey, Event extends PropertyKey>(
  spec: WorkflowSpec,
  states: readonly State[],
  events: readonly Event[],
  rows: readonly MatrixRow<State, Event>[],
  traces: readonly Trace<State, Event>[],
): void => {
  const errors = validateWorkflowSpecConformance(spec, states, events, rows, traces);
  if (errors.length > 0) {
    throw new Error(`workflow spec mismatch:\n${errors.map((err) => `  - ${err}`).join("\n")}`);
  }
};
